src/aspire/source/xform.py

```diff
@@ -0,0 +1,388 @@
+            idx = np.where(self.indices == i)[0]
+            # For the incoming Image object, find out which transformation indices are applicable
+            idx = np.intersect1d(idx, indices)
+            # For the transformation indices we found, find the indices in the Image object that we'll use
+            im_data_indices = np.where(np.isin(indices, idx))[0]
+            # Apply the transformation to the selected indices in the Image object
+            if len(im_data_indices) > 0:
+                fn_handle = getattr(self.unique_xforms[i], which)
+                im_data[:, :, im_data_indices] = fn_handle(Image(im_data[:, :, im_data_indices])).asnumpy()
+        return Image(im_data)
+
+    def _forward(self, im, indices):
+        return self._forward_or_adjoint(im, indices, 'forward')
+
+    def _adjoint(self, im, indices):
+        return self._forward_or_adjoint(im, indices, 'adjoint')
+
+    def downsample(self, resolution):
+        for xform in self.unique_xforms:
+            xform.downsample(resolution)
+        super().downsample(resolution)
+
+
+def _apply_transform(xform, im, indices, adjoint=False):
+    """
+    A simple global function (i.e. not a method) that is capable of being cached by joblib.Memory object's `cache`
+    method.
+    """
+    if not adjoint:
+        logger.info('  Applying ' + str(xform))
+        return xform.forward(im, indices=indices)
+    else:
+        logger.info('  Applying Adjoint ' + str(xform))
+        return xform.adjoint(im, indices=indices)
+
+
+class Pipeline(Xform):
+    """
+    A Pipeline is a Xform made up of individual transformation steps (i.e. multiple Xform objects).
+    The Pipeline, just like any other Xform, can be run in the `forward` or `adjoint` mode.
+
+    In addition to keeping client-side code clean, a major advantage of Pipeline is that individual steps of the pipeline
+    can be cached transparently by the Pipeline, providing significant performance advantages for steps that are performed
+    repeatedly (especially during development while setting up these pipelines) on any Image/Xform pair. This caching
+    uses joblib.Memory object behind the scenes, but is disabled by default.
+    """
+
+    def __init__(self, xforms=None, memory=None):
+        """
+        Initialize a Pipeline with Xform objects.
+        :param xforms: An iterable of Xform objects to use in the Pipeline.
+        :param memory: None for no caching (default), or the location of a directory to use to cache steps of the 
+            pipeline.
+        """
+        self.xforms = xforms or []
+        self.memory = memory
+        self.active = True
+        self._register_xforms_changed()
+
+    def _register_xforms_changed(self):
+        # Update resolution of the Pipeline in case any of the transformations changed.
+        if len(self.xforms) > 0:
+            self.resolution = min(np.inf, *[xform.resolution for xform in self.xforms])
```

Comment:
Shouldn't `Pipeline.resolution` be the resolution of the first `Xform` in the pipeline?

Reply:
So the whole reason that currently we can add downsampling as a lazy operation (done by appending a `DownSample` Xform towards the end of the generation pipeline, is because the individual `Xform` objects in a `Pipeline` don't have to be of the same resolution (i.e. they can spit out `Image` objects in different resolution).

I'll have to think about this one.

Reply:
We've now done away with a `Pipeline` maintaining it's `resolution` . Because we're not maintaining a `model_pipeline` anymore, there is no need to downsample each `Xform` in the `Pipeline`, and thus no need to maintain a `resolution` attribute for a `Xform` object (except for a `DownSample` Xform of course).